riscv-runtime/src/hash.rs

```diff
@@ -27,3 +27,85 @@ pub fn poseidon_gl(data: [u64; 12]) -> [u64; 4] {
 
     poseidon_gl_unsafe(data)
 }
+
+/// Calls the keccakf machine
+/// Return value is placed in the output array.
+/// This is unsafe because it does not check if the u64 elements fit the Goldilocks field.
+pub fn keccakf_unsafe(input: &[u64; 25], output: &mut [u64; 25]) {
+    unsafe {
+        // syscall inputs: memory pointer to input array and memory pointer to output array
+        asm!("ecall", in("a0") input as *const [u64; 25], in("a1") output as *mut [u64; 25], in("t0") u32::from(Syscall::KeccakF));
+    }
+}
+
+/// Calls the keccakf machine
+/// This function will panic if any of the u64 elements doesn't fit the Goldilocks field.
+pub fn keccakf(input: &[u64; 25], output: &mut [u64; 25]) {
+    for &n in input.iter() {
+        assert!(n < GOLDILOCKS);
+    }
```

Comment:
I don't think that's needed!

I think this code has been copied from the Poseidon implementation. But in the case of Poseidon, the data (passed as `[u64; 12]` is interpreted as 12 field elements, so we have to make sure it is in the right range (otherwise, the hash of `[0, ...]` and `[0xffffffff00000001, ...]` would be the same!).

But in the case of Keccak, the input is just seen as a block of 1600 bits (represented as an array of 50 32-Bit words, so it can't overflow the ~64-Bit Goldilocks field).

Reply:
I mean shouldn't we then just use `[u32; 50]` for the keccakf api?